### PR TITLE
allow getMarketPrice for crypto currencies

### PR DESCRIPTION
**Pull request by premek**
this makes the 'Can get market prices' API test pass

The api works for crypto currencies too but the price is inverted, e.g. for BTC it returns 250 (1 BTC = 250 XMR), for USD it returns 242 (1 XMR = 242 USD), maybe we should fix that?

### CODE REVIEW

**Comment by woodser** [comment]
Yeah I noticed there was some jankiness in how the price request was handled, something to do with 2 requests stepping on each other.

As implemented, these errors are logged to Alice's console:

```
Nov-17 10:54:59.855 [BisqDaemonMain] WARN  b.c.p.price.PriceFeedService: Exception at applyPriceToConsumer for currency USD. priceProvider=https://price.bisq.wiz.biz/. Exception=java.lang.IllegalStateException: Stream is already completed, no further calls are allowed 
Nov-17 10:54:59.855 [BisqDaemonMain] WARN  bisq.core.api.CorePriceService: Exception at applyPriceToConsumer for currency USD. priceProvider=https://price.bisq.wiz.biz/. Exception=java.lang.IllegalStateException: Stream is already completed, no further calls are allowed bisq.core.provider.price.PriceRequestException: Exception at applyPriceToConsumer for currency USD. priceProvider=https://price.bisq.wiz.biz/. Exception=java.lang.IllegalStateException: Stream is already completed, no further calls are allowed
	at bisq.core.provider.price.PriceFeedService.applyPriceToConsumer(PriceFeedService.java:375)
	at bisq.core.provider.price.PriceFeedService.lambda$request$0(PriceFeedService.java:172)
	at bisq.core.provider.price.PriceFeedService$1.lambda$onSuccess$0(PriceFeedService.java:405)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
```

I struggled to address this, eventually [reverting my PR](https://github.com/haveno-dex/haveno/commit/4d81d98e00a20756a532c3db045aa1d17fe8a674) after noticing errors.

**Review by woodser — CHANGES_REQUESTED** [review]
See my comment: https://github.com/haveno-dex/haveno/pull/165#issuecomment-971719748

**Comment by premek** [comment]
Thanks, i missed that error. I changed the CorePriceService (that seems to be used only from grpc api calls) so it doesn't use the problematic method on the priceFeedService and instead uses cached values. Cache is refreshed from here if it is older than 2 (?) minutes. 
The PriceFeedService obtained through injection should be always the same instance

File: core/src/main/java/bisq/core/api/CorePriceService.java [hunk 1]
```diff
@@ -44,26 +45,20 @@ public CorePriceService(PriceFeedService priceFeedService) {
...
-
         if (!priceFeedService.hasPrices())
             throw new IllegalStateException("price feed service has no prices");
 
-        try {
-            priceFeedService.setCurrencyCode(upperCaseCurrencyCode);
-        } catch (Throwable throwable) {
-            log.warn("Could not set currency code in PriceFeedService", throwable);
+        // fetch fresh prices if the cache is older than 2 minutes
+        if (System.currentTimeMillis() - priceFeedService.getLastRequestTimeStamp().getTime() > TimeUnit.MINUTES.toMillis(2)) {
```
Comment (woodser) [inline core/src/main/java/bisq/core/api/CorePriceService.java:52]: If I understand right, the API can serve a stale price for almost 2 minutes with this change. Also this approach merely avoids making a simultaneous request at the API level, whereas solving the underlying issue is preferred.

**Comment by premek** [comment]
We could remove the if and fetch the price every time. I just thought it's
not necessary. It downloads all currencies every time and stores them in
the cache.

I also don't know how often the price node updates the price, it wouldn't
make sense to ask more often than that.

Should we make it 30s for example? Or fetch it every time?

Why I did it like this is because it looks like the original method was
made to support something like you set your currency in the desktop gui
combo box or something to be USD for example and then you see all the
prices in the gui in USD. That's why you used setcurrency and then fetched
the prices separately. It was probably not meant to support concurrent
fetch of different currencies.

I could also try to "fix" the underlying service but it's used on a few
places in the desktop gui

**Comment by woodser** [comment]
> it wouldn't make sense to ask more often than that.

If price data updates 1 second after we refresh the cache, we would serve stale data until the next time we refresh the cache. I don't want to introduce our own lag to refresh price data, so we would want some way to avoid that.

This issue to fetch a price for a currency code is actually a small subset of this other issue to fetch all market prices, which is assigned to you: https://github.com/haveno-dex/haveno/issues/139

Maybe it makes most sense to implement that other call and then use it to power this call with a specific currency code.

**Comment by premek** [comment]
closing this, will do this as part of #139

**Pull request closed without merging**